File: ios/Runner/PgpApi.swift
//
//  PgpApi.swift
//  Runner
//
//  Created by Alexander Orlov on 26.12.2019.
//  Copyright © 2019 The Chromium Authors. All rights reserved.
//

import Foundation
import DMSOpenPGP
import BouncyCastle_ObjC

class PgpApi{
    
    let pgp = Pgp()
    var privateKey : String?=nil
    var publicKey : [String]?=nil
    var tempFile : String?=nil
    
    func setPrivateKey(_ data:String?) throws{
        privateKey=data
    }
    
    func setPublicKeys(_ data:[String]?) throws{
        
        publicKey=data
    }
    
    func decryptBytes(_ data:Data,_ password:String)throws->Data{
        let byteArray = IOSByteArray(nsData: data)!
        let input = JavaIoByteArrayInputStream(byteArray: byteArray)
        let output = JavaIoByteArrayOutputStream()
        try  decrypt(input,output,password)
        return output.toByteArray().toNSData()
    }
    
    func decryptFile(_ inputFile:String,_ outputFile:String,_ password:String)throws{
        let input = JavaIoFileInputStream(  javaIoFile: JavaIoFile(nsString: inputFile))
        let output = JavaIoFileOutputStream(  javaIoFile: JavaIoFile(nsString: outputFile))
        try decrypt(input,output,password)
    }
    
    func encryptBytes(_ data:Data,_ passwordForSign:String?)throws->Data{
        let byteArray = IOSByteArray(nsData: data)!
        let input = JavaIoByteArrayInputStream(byteArray: byteArray)
        let input2 = JavaIoByteArrayInputStream(byteArray: byteArray)
        let output = JavaIoByteArrayOutputStream()
        try  encrypt(input,input2,output,passwordForSign)
        return output.toByteArray().toNSData()
    }
    
    func encryptFile(_ inputFile:String,_ outputFile:String,_ passwordForSign:String?)throws{
        let input = JavaIoFileInputStream(  javaIoFile: JavaIoFile(nsString: inputFile))
        let input2 = JavaIoFileInputStream(  javaIoFile: JavaIoFile(nsString: inputFile))
        let output = JavaIoFileOutputStream(  javaIoFile: JavaIoFile(nsString: outputFile))
        try encrypt(input,input2,output,passwordForSign)
    }
    
    func encrypt(_ input:JavaIoInputStream,_ input2:JavaIoInputStream,_ output:JavaIoOutputStream,_ passwordForSign:String?) throws {
        assert(publicKey != nil)
        if(passwordForSign != nil){
            assert(privateKey != nil)
        }
        try pgp.encrypt(input,input2,output,publicKey!,privateKey,passwordForSign)
    }
    func decrypt(_ input:JavaIoInputStream,_ output:JavaIoOutputStream,_ password:String) throws {
        assert(privateKey != nil)
        try pgp.decrypt(input,output,privateKey!,password)
    }
    func getKeyDescription(_ key:Data) throws->KeyInfo{
        return try pgp.getKeyDescription(key)
    }
    func createKeys(_ length:Int32,_ email:String,_ password:String) throws->[String]{
        return try pgp.createKeys(length,email,password)
    }
    
    func setTempFile(_ file:String?){
        tempFile=file
        
    }
    
    func encryptSymmetricBytes(_ data:Data,_ password:String)throws ->Data {
        let byteArray = IOSByteArray(nsData: data)!
        let input = JavaIoByteArrayInputStream(byteArray: byteArray)
        let output = JavaIoByteArrayOutputStream()
        try encryptSymmetric(input,output,password,jlong(data.count))
        return output.toByteArray().toNSData()
    }
    func encryptSymmetricFile(_ inputFile:String,_ outputFile:String,_ password:String) throws {
        let file=JavaIoFile(nsString: inputFile)
        let input = JavaIoFileInputStream(  javaIoFile: file)
        let output = JavaIoFileOutputStream(  javaIoFile: JavaIoFile(nsString: outputFile))
        try encryptSymmetric(input,output,password,file.length())
    }
    func encryptSymmetric(
        _ input:JavaIoInputStream,
        _ output:JavaIoOutputStream,
        _ password:String,
        _ length:jlong
    ) throws{
        pgp.encryptSymetric(input,output,password,length)
    }
    
    func decryptSymmetricBytes(_ data:Data,_ password:String) throws->Data{
        let byteArray = IOSByteArray(nsData: data)!
        let input = JavaIoByteArrayInputStream(byteArray: byteArray)
        let output = JavaIoByteArrayOutputStream()
        try decryptSymmetric(input,output,password,jlong(data.count))
        return output.toByteArray().toNSData()
    }
    func decryptSymmetricFile(_ inputFile:String,_ outputFile:String,_ password:String) throws{
        let file=JavaIoFile(nsString: inputFile)
        let input = JavaIoFileInputStream(  javaIoFile: file)
        let output = JavaIoFileOutputStream(  javaIoFile: JavaIoFile(nsString: outputFile))
        try decryptSymmetric(input,output,password,file.length())
    }
    func decryptSymmetric(
        _ input:JavaIoInputStream,
        _ output:JavaIoOutputStream,
        _ password:String,
        _ length:Int64
    ) throws {
        pgp.decryptSymmetric(input,output,password,length)
    }
    
    func checkPassword(_ password:String,_ privateKey:String)throws -> Bool {
        return pgp.checkPassword(password,privateKey)
    }
    
}

File: core/crypto_plugin/android/src/main/kotlin/com/afterlogic/crypto_plugin/pgp/PgpApi.kt
package com.afterlogic.crypto_plugin.pgp

import KeyDescription
import java.io.*

open class PgpApi {
    private val pgp = Pgp()
    private var publicKey: List<String>? = null
    private var privateKey: String? = null
    private var tempFile: File? = null

    fun getKeyDescription(key: String): KeyDescription {
        return pgp.getEmailFromKey(ByteArrayInputStream(key.toByteArray()))
    }

    fun setPrivateKey(key: String?) {
        privateKey = key
    }

    fun setPublicKeys(key: List<String>?) {
        publicKey = key
    }

    fun setTempFile(filePath: String?) {
        tempFile =
                filePath?.let {
                    File(it).apply {
                        createNewFile()
                        assert(isFile)
                        assert(canWrite())
                    }
                }

    }

    fun getProgress(): Progress? {
        return if (pgp.progress?.complete == false) {
            pgp.progress
        } else {
            null
        }
    }

    private fun decrypt(inputStream: InputStream, outputStream: OutputStream, password: String?, length: Long, checkSign: Boolean) {
        if (checkSign) {
            assert(publicKey != null)
        }
        pgp.decrypt(inputStream, outputStream, privateKey, password, length, if (checkSign) {
            publicKey!!.firstOrNull()
        } else {
            null
        }
        )

    }

    fun decryptBytes(array: ByteArray, password: String?, checkSign: Boolean = false): ByteArray {
        val outStream = ByteArrayOutputStream()
        decrypt(ByteArrayInputStream(array), outStream, password, array.size.toLong(), checkSign)
        return outStream.toByteArray()
    }


    fun decryptFile(inputFilePath: String, outputFilePath: String, password: String?, checkSign: Boolean = false) {
        val inputFile = File(inputFilePath)
        val outputFile = File(outputFilePath)
        assert(inputFile.isFile)
        assert(inputFile.exists())
        outputFile.createNewFile()
        assert(outputFile.isFile)
        assert(outputFile.canWrite())
        decrypt(FileInputStream(inputFile), FileOutputStream(outputFile), password, inputFile.length(), checkSign)

    }

    private fun encrypt(outputStream: OutputStream, inputStream: InputStream, length: Long, password: String?) {
        val sign = password != null
        if (sign) {
            assert(privateKey != null)
        }
        pgp.encrypt(
                outputStream,
                inputStream,
                publicKey,
                if (sign) {
                    privateKey
                } else {
                    null
                },
                password,
                length
        )
    }

    fun encriptFile(inputFilePath: String, outputFilePath: String, password: String?) {
        val inputFile = File(inputFilePath)
        val outputFile = File(outputFilePath)
        assert(inputFile.isFile)
        assert(inputFile.exists())
        assert(inputFile.canRead())
        outputFile.createNewFile()
        assert(outputFile.isFile)
        assert(outputFile.canWrite())
        encrypt(FileOutputStream(outputFile), FileInputStream(inputFile), inputFile.length(), password)
    }

    fun encriptBytes(array: ByteArray, password: String?): ByteArray {
        val outStream = ByteArrayOutputStream()
        encrypt(outStream, ByteArrayInputStream(array), array.size.toLong(), password)
        return outStream.toByteArray()
    }

    fun createKeys(length: Int, email: String, password: String): List<String> {
        return pgp.createKeys(length, email, password).mapIndexed { i, it ->
            it.toString(Charsets.UTF_8)
        }
    }


    fun decryptSymmetric(inputStream: InputStream, outputStream: OutputStream, password: String, sign: Boolean) {
        pgp.symmetricallyDecrypt(inputStream, outputStream, password, if (sign) {
            assert(publicKey != null)
            publicKey!!.firstOrNull()
        } else {
            null
        })
    }

    fun encryptSymmetric(inputStream: InputStream, outputStream: OutputStream, length: Long, password: String, keyPassword: String?) {
        assert(tempFile != null)
        pgp.symmetricallyEncrypt(
                inputStream,
                outputStream,
                tempFile!!,
                length,
                password,
                if (keyPassword != null) {
                    assert(privateKey != null)
                    privateKey
                } else {
                    null
                },
                keyPassword
        )
    }

    fun encryptSymmetricBytes(array: ByteArray, password: String, keyPassword: String? = null): ByteArray {
        val outStream = ByteArrayOutputStream()
        encryptSymmetric(ByteArrayInputStream(array), outStream, array.size.toLong(), password, keyPassword)
        return outStream.toByteArray()
    }

    fun encryptSymmetricFile(inputFilePath: String, outputFilePath: String, password: String, keyPassword: String? = null) {
        val inputFile = File(inputFilePath)
        val outputFile = File(outputFilePath)
        assert(inputFile.isFile)
        assert(inputFile.exists())
        assert(inputFile.canRead())
        outputFile.createNewFile()
        assert(outputFile.isFile)
        assert(outputFile.canWrite())
        encryptSymmetric(FileInputStream(inputFile), FileOutputStream(outputFile), inputFile.length(), password, keyPassword)
    }

    fun decryptSymmetricBytes(array: ByteArray, password: String, checkSign: Boolean = false): ByteArray {
        val outStream = ByteArrayOutputStream()
        decryptSymmetric(ByteArrayInputStream(array), outStream, password, checkSign)
        return outStream.toByteArray()
    }

    fun decryptSymmetricFile(inputFilePath: String, outputFilePath: String, password: String, checkSign: Boolean = false) {
        val inputFile = File(inputFilePath)
        val outputFile = File(outputFilePath)
        assert(inputFile.isFile)
        assert(inputFile.exists())
        outputFile.createNewFile()
        assert(outputFile.isFile)
        assert(outputFile.canWrite())
        decryptSymmetric(FileInputStream(inputFile), FileOutputStream(outputFile), password, checkSign)
    }

    fun checkPassword(password: String, privateKey: String): Boolean {
        return pgp.checkPassword(password, privateKey)
    }
}